File: typescript/src/mpc_designation_cli.ts
#!/usr/bin/env node
/**
 * Command-line interface for MPC designation conversion.
 */

import { convert, VERSION, MPCDesignationError } from './mpc_designation';

function printUsage(): void {
    console.log('Usage: mpc_designation [options] <designation>');
    console.log();
    console.log('Convert between packed and unpacked MPC designations.');
    console.log();
    console.log('Options:');
    console.log('  -v, --verbose   Show detailed format information');
    console.log('  -h, --help      Show this help message');
    console.log('  --version       Show version');
    console.log();
    console.log('Examples:');
    console.log("  mpc_designation '1995 XA'     # Convert to packed: J95X00A");
    console.log('  mpc_designation J95X00A       # Convert to unpacked: 1995 XA');
    console.log("  mpc_designation -v '1P'       # Show verbose output");
}

function main(): void {
    const args = process.argv.slice(2);

    if (args.length === 0) {
        printUsage();
        process.exit(1);
    }

    let verbose = false;
    let designation: string | null = null;

    for (let i = 0; i < args.length; i++) {
        const arg = args[i];
        if (arg === '-h' || arg === '--help') {
            printUsage();
            process.exit(0);
        } else if (arg === '--version') {
            console.log(`mpc_designation ${VERSION}`);
            process.exit(0);
        } else if (arg === '-v' || arg === '--verbose') {
            verbose = true;
        } else {
            // Collect remaining args as the designation
            designation = args.slice(i).join(' ');
            break;
        }
    }

    if (!designation) {
        console.error('Error: No designation provided');
        printUsage();
        process.exit(1);
    }

    try {
        const result = convert(designation);

        if (verbose) {
            console.log(`Input:   ${result.input}`);
            console.log(`Output:  ${result.output}`);
            console.log(`Format:  ${result.info.format}`);
            console.log(`Type:    ${result.info.type}`);
            console.log(`Subtype: ${result.info.subtype}`);
        } else {
            console.log(result.output);
        }
    } catch (e) {
        if (e instanceof MPCDesignationError) {
            console.error(`Error: ${e.message}`);
        } else {
            console.error(`Error: ${e}`);
        }
        process.exit(1);
    }
}

main();

File: fortran/src/mpc_designation_cli.f90
!> Command-line interface for MPC designation conversion
program mpc_designation_cli
    use mpc_designation
    implicit none

    character(len=200) :: arg, designation
    type(conversion_result) :: res
    integer :: i, nargs
    logical :: verbose

    verbose = .false.
    designation = ''
    nargs = command_argument_count()

    if (nargs == 0) then
        call print_usage()
        stop 1
    end if

    i = 1
    do while (i <= nargs)
        call get_command_argument(i, arg)

        if (arg == '-h' .or. arg == '--help') then
            call print_usage()
            stop 0
        else if (arg == '--version') then
            print '(A)', 'mpc_designation ' // VERSION
            stop 0
        else if (arg == '-v' .or. arg == '--verbose') then
            verbose = .true.
        else
            ! Collect remaining args as designation
            designation = trim(arg)
            do while (i < nargs)
                i = i + 1
                call get_command_argument(i, arg)
                designation = trim(designation) // ' ' // trim(arg)
            end do
            exit
        end if

        i = i + 1
    end do

    if (len_trim(designation) == 0) then
        print '(A)', 'Error: No designation provided'
        call print_usage()
        stop 1
    end if

    res = convert(designation)

    if (MPCDesignationError) then
        print '(A)', 'Error: ' // trim(error_message)
        stop 1
    end if

    if (verbose) then
        print '(A)', 'Input:   ' // trim(res%input)
        print '(A)', 'Output:  ' // trim(res%output)
        print '(A)', 'Format:  ' // trim(res%info%format)
        print '(A)', 'Type:    ' // trim(res%info%dtype)
        print '(A)', 'Subtype: ' // trim(res%info%subtype)
    else
        print '(A)', trim(res%output)
    end if

contains

    subroutine print_usage()
        print '(A)', 'Usage: mpc_designation [options] <designation>'
        print '(A)', ''
        print '(A)', 'Convert between packed and unpacked MPC designations.'
        print '(A)', ''
        print '(A)', 'Options:'
        print '(A)', '  -v, --verbose   Show detailed format information'
        print '(A)', '  -h, --help      Show this help message'
        print '(A)', '  --version       Show version'
        print '(A)', ''
        print '(A)', 'Examples:'
        print '(A)', "  mpc_designation '1995 XA'     # Convert to packed: J95X00A"
        print '(A)', '  mpc_designation J95X00A       # Convert to unpacked: 1995 XA'
    end subroutine print_usage

end program mpc_designation_cli
